File: exercises/domain-2/block-2.2/solution/solution.py
import asyncio
import random
from dataclasses import dataclass


@dataclass
class ToolResult:
    success: bool
    data: dict | None = None
    error: str | None = None
    retryable: bool | None = None
    suggestion: str | None = None


@dataclass
class FetchResult:
    status: int
    body: str


async def simulate_fetch(url: str) -> FetchResult:
    rand = random.random()
    if rand < 0.3:
        raise ConnectionError("ECONNREFUSED")
    if rand < 0.5:
        return FetchResult(503, "Service Unavailable")
    if "missing" in url:
        return FetchResult(404, "Not Found")
    return FetchResult(200, f"<html>Content of {url}</html>")


async def robust_fetch_tool(url: str) -> ToolResult:
    MAX_RETRIES = 3

    for attempt in range(MAX_RETRIES):
        try:
            result = await simulate_fetch(url)

            if 200 <= result.status < 300:
                return ToolResult(success=True, data={"status": result.status, "body": result.body})

            if 400 <= result.status < 500:
                return ToolResult(
                    success=False,
                    error=f"HTTP {result.status}: {result.body}",
                    retryable=False,
                    suggestion="Verify the URL is correct." if result.status == 404 else "Check request parameters.",
                )

            # 5xx — retry
            if attempt < MAX_RETRIES - 1:
                print(f"  Attempt {attempt + 1} failed ({result.status}), retrying...")
                await asyncio.sleep(1.0 * (2 ** attempt))
                continue

            return ToolResult(
                success=False,
                error=f"Server error after {MAX_RETRIES} attempts: HTTP {result.status}",
                retryable=True,
                suggestion="The server is experiencing issues. Try again later.",
            )

        except Exception as e:
            if attempt < MAX_RETRIES - 1:
                print(f"  Attempt {attempt + 1} network error, retrying...")
                await asyncio.sleep(1.0 * (2 ** attempt))
                continue

            return ToolResult(
                success=False,
                error=f"Network error after {MAX_RETRIES} attempts: {e}",
                retryable=True,
                suggestion="Check network connectivity.",
            )

    return ToolResult(success=False, error="Max retries exceeded", retryable=True)


async def main():
    print("Testing valid URL:")
    print(await robust_fetch_tool("https://example.com/page"))
    print("\nTesting missing URL:")
    print(await robust_fetch_tool("https://example.com/missing"))


asyncio.run(main())

File: exercises/domain-2/block-2.2/solution/solution.ts
interface ToolResult<T = unknown> {
  success: boolean
  data?: T
  error?: string
  retryable?: boolean
  suggestion?: string
}

interface FetchResult {
  status: number
  body: string
}

async function simulateFetch(url: string): Promise<FetchResult> {
  const rand = Math.random()
  if (rand < 0.3) throw new Error("ECONNREFUSED: connection refused")
  if (rand < 0.5) return { status: 503, body: "Service Unavailable" }
  if (url.includes("missing")) return { status: 404, body: "Not Found" }
  return { status: 200, body: `<html>Content of ${url}</html>` }
}

function sleep(ms: number) { return new Promise(resolve => setTimeout(resolve, ms)) }

async function robustFetchTool(url: string): Promise<ToolResult<FetchResult>> {
  const MAX_RETRIES = 3

  for (let attempt = 0; attempt < MAX_RETRIES; attempt++) {
    try {
      const result = await simulateFetch(url)

      if (result.status >= 200 && result.status < 300) {
        return { success: true, data: result }
      }

      if (result.status >= 400 && result.status < 500) {
        // Client error — don't retry
        return {
          success: false,
          error: `HTTP ${result.status}: ${result.body}`,
          retryable: false,
          suggestion: result.status === 404
            ? "The URL does not exist. Verify the URL is correct with list_available_pages."
            : "Check request parameters and try a different URL.",
        }
      }

      // 5xx — transient, retry
      if (attempt < MAX_RETRIES - 1) {
        console.log(`  Attempt ${attempt + 1} failed (${result.status}), retrying...`)
        await sleep(1000 * Math.pow(2, attempt))
        continue
      }

      return {
        success: false,
        error: `Server error after ${MAX_RETRIES} attempts: HTTP ${result.status}`,
        retryable: true,
        suggestion: "The server appears to be experiencing issues. Try again later.",
      }

    } catch (err) {
      // Network error — transient, retry
      if (attempt < MAX_RETRIES - 1) {
        console.log(`  Attempt ${attempt + 1} network error, retrying...`)
        await sleep(1000 * Math.pow(2, attempt))
        continue
      }

      return {
        success: false,
        error: `Network error after ${MAX_RETRIES} attempts: ${err}`,
        retryable: true,
        suggestion: "Check network connectivity and try again.",
      }
    }
  }

  return { success: false, error: "Max retries exceeded", retryable: true }
}

async function main() {
  console.log("Testing with valid URL:")
  console.log(await robustFetchTool("https://example.com/page"))
  console.log("\nTesting with missing URL:")
  console.log(await robustFetchTool("https://example.com/missing"))
}

main()
